app/src/heartbeat/heartbeat_controller_desktop.cc

```diff
@@ -79,5 +97,171 @@ void HeartbeatController::LogHeartbeat() {
+      // If read fails, or if there are no stored heartbeats, return an empty
+      // payload and don't attempt to write.
+      if (read_succeeded && logged_heartbeats.heartbeats.size() > 0) {
+        // Clear all logged heartbeats, but keep the last logged date
+        LoggedHeartbeats cleared_heartbeats;
+        cleared_heartbeats.last_logged_date =
+            logged_heartbeats.last_logged_date;
+        bool write_succeeded = this->storage_.Write(cleared_heartbeats);
+        // Only update last-logged date and return a payload if the write
+        // succeeds.
+        if (write_succeeded) {
+          this->last_fetched_all_heartbeats_date_ = current_date;
+          *output_str = GetStringPayloadForHeartbeats(logged_heartbeats);
+        }
+      }
+    }
+    // Post the semaphore to signal the main thread to read the result.
+    scheduled_work_semaphore->Post();
+    return;
+  };
+
+  scheduler_.Schedule(get_and_reset_function);
+  // Wait until the scheduled work completes.
+  if (scheduled_work_semaphore->TimedWait(kMaxWaitTimeMs)) {
+    return *output_str;
+  }
+  // Return an empty string if unable to TimedWait times out.
+  return "";
+}
+
+std::string HeartbeatController::GetAndResetTodaysStoredHeartbeats() {
+  std::shared_ptr<std::string> output_str = std::make_shared<std::string>("");
+  std::shared_ptr<Semaphore> scheduled_work_semaphore =
+      std::make_shared<Semaphore>(0);
+
+  std::function<void(void)> get_and_reset_function = [&, output_str]() {
+    std::string current_date = date_provider_.GetDate();
+    // Return early if a heartbeat has already been logged today.
+    if (this->last_fetched_all_heartbeats_date_ != current_date) {
+      LoggedHeartbeats stored_heartbeats;
+      bool read_succeeded = this->storage_.ReadTo(stored_heartbeats);
+      // If read fails, or if there are no stored heartbeats, return an empty
+      // payload and don't attempt to write.
+      if (read_succeeded && stored_heartbeats.heartbeats.size() > 0) {
+        // Find a logged heartbeat from today. Remove it from the stored
+        // heartbeats and use it to construct a single-heartbeat payload.
+        LoggedHeartbeats todays_heartbeats;
+        for (auto& entry : stored_heartbeats.heartbeats) {
+          std::string user_agent = entry.first;
+          std::vector<std::string>& dates = entry.second;
+          auto itr = std::find(dates.begin(), dates.end(), current_date);
+          if (itr != dates.end()) {
+            dates.erase(itr);
+            todays_heartbeats.heartbeats[user_agent].push_back(current_date);
+          }
+        }
+        bool write_succeeded = this->storage_.Write(stored_heartbeats);
+        // Only update last-logged date and return a payload if the write
+        // succeeds.
+        if (write_succeeded) {
+          this->last_fetched_todays_heartbeat_date_ = current_date;
+          *output_str = GetStringPayloadForHeartbeats(todays_heartbeats);
```

Comment:
I think the idea is that the v1 calling code will just check whether or not the result is entry. I could change this to return a boolean, but it was simpler to duplicate the code and return the same payload.
I'll look a bit more closely into how this would be used.

Reply:
To be more specific:
This seems to be the relevant iOS code which just checks whether or not the payload is empty:
https://source.corp.google.com/github/firebase/firebase-ios-sdk/FirebaseCore/Sources/FIRHeartbeatLogger.m;rcl=6472d7c8e0e974d5a78370353e58226d1bb07ca1;l=81

This is the relevant C++ code that we are trying to replace for v1 heartbeats:
https://source.corp.google.com/piper///depot/google3/third_party/firebase/cpp/app/src/heartbeat_info_desktop.cc;l=83;rcl=463226221
This c++ code basically just compares a timestamp and decides whether to return heartbeat code SDK, global, or None.

So For the new method to get todays heartbeat, I think it could return a payload string, or even just return a boolean. The contents of the string don't seem to matter at all.
Which would you prefer?